File: volleyball_stats/app/logic/segment_rules.py
"""
Модуль правил для сегментов действий (VolleyActionSegmentRules)
Определяет, какие действия могут следовать за другими, и результаты действий
Аналог StatisticsCreatorModule/Logic/VolleyActionSegmetRules.cs
"""
from typing import List, Dict
from app.models.action import VolleyActionType
from app.models.sequence import ActionSegmentResult


class VolleyActionSegmentRules:
    """
    Правила для последовательности действий в волейболе
    Определяет:
    1. Какие действия могут следовать после данного действия с определенным качеством
    2. Результат сегмента (выигрыш/проигрыш/не закончен) на основе действия и качества
    """
    
    def __init__(self, default_settings: bool = True):
        # sequence_logic[action_type][quality-1] = list of possible next actions
        self._sequence_logic: Dict[VolleyActionType, List[List[VolleyActionType]]] = {}
        
        # win_loss_logic[action_type][quality-1] = ActionSegmentResult
        self._win_loss_logic: Dict[VolleyActionType, List[ActionSegmentResult]] = {}
        
        if default_settings:
            self.set_default_rules()
    
    def set_rule_for_action_and_quality(self, 
                                        act_type: VolleyActionType, 
                                        quality: int, 
                                        *types: VolleyActionType) -> None:
        """
        Устанавливает правило: после действия act_type с качеством quality
        могут следовать действия types
        """
        if act_type not in self._sequence_logic:
            self._sequence_logic[act_type] = [[] for _ in range(6)]
        
        self._sequence_logic[act_type][quality - 1] = list(types)
    
    def set_rule_for_action_and_qualities(self,
                                         act_type: VolleyActionType,
                                         qualities: List[int],
                                         *types: VolleyActionType) -> None:
        """Устанавливает правило для нескольких качеств сразу"""
        for qual in qualities:
            self.set_rule_for_action_and_quality(act_type, qual, *types)
    
    def set_win_lose_rule(self, act_type: VolleyActionType, results: List[ActionSegmentResult]) -> None:
        """Устанавливает правило выигрыша/проигрыша для действия"""
        self._win_loss_logic[act_type] = results
    
    def set_default_rules(self) -> None:
        """Устанавливает стандартные правила волейбола"""
        
        # SERVE (Подача)
        self.set_rule_for_action_and_quality(
            VolleyActionType.SERVE, 6,  # Эйс
            VolleyActionType.SERVE
        )
        self.set_rule_for_action_and_qualities(
            VolleyActionType.SERVE, [5, 4, 3, 2],  # Хорошая подача
            VolleyActionType.BLOCK, VolleyActionType.DEFENCE, 
            VolleyActionType.FREE_BALL, VolleyActionType.SET, VolleyActionType.ATTACK
        )
        self.set_rule_for_action_and_quality(
            VolleyActionType.SERVE, 1,  # Ошибка подачи
            VolleyActionType.RECEPTION
        )
        
        # RECEPTION (Прием)
        self.set_rule_for_action_and_qualities(
            VolleyActionType.RECEPTION, [6, 5, 4, 3],  # Хороший прием
            VolleyActionType.SET, VolleyActionType.ATTACK, 
            VolleyActionType.BLOCK, VolleyActionType.DEFENCE, VolleyActionType.FREE_BALL
        )
        self.set_rule_for_action_and_quality(
            VolleyActionType.RECEPTION, 2,  # Плохой прием
            VolleyActionType.BLOCK, VolleyActionType.DEFENCE, 
            VolleyActionType.FREE_BALL, VolleyActionType.SET, VolleyActionType.ATTACK
        )
        self.set_rule_for_action_and_quality(
            VolleyActionType.RECEPTION, 1,  # Ошибка приема
            VolleyActionType.RECEPTION
        )
        
        # SET (Передача)
        self.set_rule_for_action_and_qualities(
            VolleyActionType.SET, [6, 5, 4, 3, 2],
            VolleyActionType.ATTACK, VolleyActionType.TRANSFER,
            VolleyActionType.BLOCK, VolleyActionType.DEFENCE, VolleyActionType.FREE_BALL
        )
        self.set_rule_for_action_and_quality(
            VolleyActionType.SET, 1,  # Ошибка передачи
            VolleyActionType.RECEPTION
        )
        
        # ATTACK (Атака)
        self.set_rule_for_action_and_quality(
            VolleyActionType.ATTACK, 6,  # Убойная атака
            VolleyActionType.SERVE
        )
        self.set_rule_for_action_and_qualities(
            VolleyActionType.ATTACK, [5, 4, 3],  # Атака в игру
            VolleyActionType.BLOCK, VolleyActionType.DEFENCE, 
            VolleyActionType.FREE_BALL, VolleyActionType.SET, VolleyActionType.ATTACK
        )
        self.set_rule_for_action_and_qualities(
            VolleyActionType.ATTACK, [2, 1],  # Ошибка атаки
            VolleyActionType.RECEPTION
        )
        
        # BLOCK (Блок)
        self.set_rule_for_action_and_quality(
            VolleyActionType.BLOCK, 6,  # Результативный блок
            VolleyActionType.SERVE
        )
        self.set_rule_for_action_and_qualities(
            VolleyActionType.BLOCK, [5, 4, 3, 2],  # Блок в игру
            VolleyActionType.BLOCK, VolleyActionType.DEFENCE, 
            VolleyActionType.FREE_BALL, VolleyActionType.SET, VolleyActionType.ATTACK
        )
        self.set_rule_for_action_and_quality(
            VolleyActionType.BLOCK, 1,  # Ошибка блока
            VolleyActionType.RECEPTION
        )
        
        # DEFENCE (Защита)
        self.set_rule_for_action_and_qualities(
            VolleyActionType.DEFENCE, [6, 5, 4, 3, 2],
            VolleyActionType.BLOCK, VolleyActionType.DEFENCE, 
            VolleyActionType.SET, VolleyActionType.ATTACK, VolleyActionType.TRANSFER
        )
        self.set_rule_for_action_and_quality(
            VolleyActionType.DEFENCE, 1,  # Ошибка защиты
            VolleyActionType.RECEPTION
        )
        
        # FREE_BALL (Свободный мяч)
        self.set_rule_for_action_and_qualities(
            VolleyActionType.FREE_BALL, [6, 5, 4, 3, 2],
            VolleyActionType.BLOCK, VolleyActionType.DEFENCE, 
            VolleyActionType.SET, VolleyActionType.ATTACK, VolleyActionType.TRANSFER
        )
        self.set_rule_for_action_and_quality(
            VolleyActionType.FREE_BALL, 1,
            VolleyActionType.RECEPTION
        )
        
        # TRANSFER (Перевод)
        self.set_rule_for_action_and_qualities(
            VolleyActionType.TRANSFER, [6, 5, 4, 3, 2],
            VolleyActionType.BLOCK, VolleyActionType.DEFENCE, 
            VolleyActionType.FREE_BALL, VolleyActionType.SET, VolleyActionType.ATTACK
        )
        self.set_rule_for_action_and_quality(
            VolleyActionType.TRANSFER, 1,
            VolleyActionType.RECEPTION
        )
        
        # OPPONENT ACTIONS (Действия соперника)
        self.set_rule_for_action_and_qualities(
            VolleyActionType.OPPONENT_POINT, [1, 2, 3, 4, 5, 6],
            VolleyActionType.RECEPTION
        )
        self.set_rule_for_action_and_qualities(
            VolleyActionType.OPPONENT_ERROR, [1, 2, 3, 4, 5, 6],
            VolleyActionType.SERVE
        )
        
        # Устанавливаем правила выигрыша/проигрыша
        self._set_default_win_lose_logic()
    
    def _set_default_win_lose_logic(self) -> None:
        """Устанавливает стандартные правила определения результата сегмента"""
        NE = ActionSegmentResult.NOT_ENDED
        WON = ActionSegmentResult.WON
        LOST = ActionSegmentResult.LOST
        
        self.set_win_lose_rule(VolleyActionType.SERVE, 
                              [LOST, NE, NE, NE, NE, WON])
        
        self.set_win_lose_rule(VolleyActionType.RECEPTION, 
                              [LOST, NE, NE, NE, NE, NE])
        
        self.set_win_lose_rule(VolleyActionType.SET, 
                              [LOST, NE, NE, NE, NE, NE])
        
        self.set_win_lose_rule(VolleyActionType.ATTACK, 
                              [LOST, LOST, NE, NE, NE, WON])
        
        self.set_win_lose_rule(VolleyActionType.BLOCK, 
                              [LOST, NE, NE, NE, NE, WON])
        
        self.set_win_lose_rule(VolleyActionType.DEFENCE, 
                              [LOST, NE, NE, NE, NE, NE])
        
        self.set_win_lose_rule(VolleyActionType.FREE_BALL, 
                              [LOST, NE, NE, NE, NE, NE])
        
        self.set_win_lose_rule(VolleyActionType.TRANSFER, 
                              [LOST, NE, NE, NE, NE, NE])
        
        self.set_win_lose_rule(VolleyActionType.OPPONENT_ERROR, 
                              [WON, WON, WON, WON, WON, WON])
        
        self.set_win_lose_rule(VolleyActionType.OPPONENT_POINT, 
                              [LOST, LOST, LOST, LOST, LOST, LOST])
    
    # ===== ИНТЕРФЕЙС ПОЛУЧЕНИЯ ПРАВИЛ =====
    
    def get_possible_actions(self, action_type: VolleyActionType, quality: int) -> List[VolleyActionType]:
        """Возвращает возможные действия после данного действия с указанным качеством"""
        if action_type in self._sequence_logic:
            return self._sequence_logic[action_type][quality - 1]
        return []
    
    def get_possible_actions_for_opponent(self, action_type: VolleyActionType) -> List[VolleyActionType]:
        """Возвращает возможные действия соперника (качество = 1)"""
        return self.get_possible_actions(action_type, 1)
    
    def get_action_result(self, action_type: VolleyActionType, quality: int) -> ActionSegmentResult:
        """Возвращает результат сегмента на основе действия и качества"""
        if action_type in self._win_loss_logic:
            return self._win_loss_logic[action_type][quality - 1]
        return ActionSegmentResult.NOT_ENDED
    
    def get_action_result_for_opponent(self, action_type: VolleyActionType) -> ActionSegmentResult:
        """Возвращает результат для действия соперника"""
        if action_type == VolleyActionType.OPPONENT_ERROR:
            return ActionSegmentResult.WON
        return ActionSegmentResult.LOST
    
    def get_action_result_for_judge(self, action_type: VolleyActionType) -> ActionSegmentResult:
        """Возвращает результат для действия судьи"""
        if action_type == VolleyActionType.DISPUTABLE_BALL:
            return ActionSegmentResult.DISPUTABLE
        if action_type == VolleyActionType.JUDGE_MISTAKE_LOST:
            return ActionSegmentResult.LOST
        return ActionSegmentResult.WON
    
    def get_set_start_players_actions(self) -> List[VolleyActionType]:
        """Возвращает действия игроков в начале сета"""
        return [VolleyActionType.SERVE, VolleyActionType.RECEPTION]
    
    def get_reception_players_actions(self) -> List[VolleyActionType]:
        """Возвращает действия приема"""
        return [VolleyActionType.RECEPTION]

File: volleyball_stats/static/js/segment-rules.js
/**
 * VolleyActionSegmentRules - правила последовательности волейбольных действий
 * Портировано из C# StatisticsCreatorModule/Logic/VolleyActionSegmetRules.cs
 * 
 * Определяет:
 * 1. Какие действия могут следовать после действия с определённым качеством
 * 2. Результат сегмента (Won/Lost/NotEnded) в зависимости от действия и качества
 */

export class VolleyActionSegmentRules {
    constructor() {
        // Правила последовательности: actionType -> [quality 1..6] -> allowed next actions
        this.sequenceLogic = {};
        
        // Правила результата: actionType -> [quality 1..6] -> segment result
        this.winLossLogic = {};
        
        this.setDefaultRules();
    }

    /**
     * Установить правило для действия и качества
     */
    setRuleForActionAndQuality(actionType, qualities, ...allowedNextActions) {
        if (!this.sequenceLogic[actionType]) {
            this.sequenceLogic[actionType] = Array(6).fill(null);
        }
        
        const qualityArray = Array.isArray(qualities) ? qualities : [qualities];
        
        qualityArray.forEach(quality => {
            this.sequenceLogic[actionType][quality - 1] = [...allowedNextActions];
        });
    }

    /**
     * Установить правило результата для действия
     */
    setWinLoseRule(actionType, results) {
        this.winLossLogic[actionType] = results;
    }

    /**
     * Установить правила по умолчанию из C# кода
     */
    setDefaultRules() {
        // SERVE - подача
        this.setRuleForActionAndQuality('Serve', 6, 'Serve'); // Ace - следующая подача
        this.setRuleForActionAndQuality('Serve', [5, 4, 3, 2], 'Block', 'Defence', 'FreeBall', 'Set', 'Attack');
        this.setRuleForActionAndQuality('Serve', 1, 'Reception'); // Ошибка - приём соперника
        
        // RECEPTION - приём
        this.setRuleForActionAndQuality('Reception', [6, 5, 4, 3], 'Set', 'Attack', 'Block', 'Defence', 'FreeBall');
        this.setRuleForActionAndQuality('Reception', 2, 'Block', 'Defence', 'FreeBall', 'Set', 'Attack');
        this.setRuleForActionAndQuality('Reception', 1, 'Reception'); // Ошибка - приём соперника
        
        // SET - передача
        this.setRuleForActionAndQuality('Set', [6, 5, 4, 3, 2], 'Attack', 'Transfer', 'Block', 'Defence', 'FreeBall');
        this.setRuleForActionAndQuality('Set', 1, 'Reception'); // Ошибка
        
        // ATTACK - атака
        this.setRuleForActionAndQuality('Attack', 6, 'Serve'); // Выигрыш - наша подача
        this.setRuleForActionAndQuality('Attack', [5, 4, 3], 'Block', 'Defence', 'FreeBall', 'Set', 'Attack');
        this.setRuleForActionAndQuality('Attack', [2, 1], 'Reception'); // Ошибка или в аут
        
        // BLOCK - блок
        this.setRuleForActionAndQuality('Block', 6, 'Serve'); // Выигрыш
        this.setRuleForActionAndQuality('Block', [5, 4, 3, 2], 'Block', 'Defence', 'FreeBall', 'Set', 'Attack');
        this.setRuleForActionAndQuality('Block', 1, 'Reception'); // Ошибка
        
        // DEFENCE - защита
        this.setRuleForActionAndQuality('Defence', [6, 5, 4, 3, 2], 'Block', 'Defence', 'Set', 'Attack', 'Transfer');
        this.setRuleForActionAndQuality('Defence', 1, 'Reception'); // Ошибка
        
        // FREEBALL - свободный мяч
        this.setRuleForActionAndQuality('FreeBall', [6, 5, 4, 3, 2], 'Block', 'Defence', 'Set', 'Attack', 'Transfer');
        this.setRuleForActionAndQuality('FreeBall', 1, 'Reception'); // Ошибка
        
        // TRANSFER - обводка
        this.setRuleForActionAndQuality('Transfer', [6, 5, 4, 3, 2], 'Block', 'Defence', 'FreeBall', 'Set', 'Attack');
        this.setRuleForActionAndQuality('Transfer', 1, 'Reception'); // Ошибка
        
        // OPPONENT ACTIONS - действия соперника
        this.setRuleForActionAndQuality('OpponentPoint', [1, 2, 3, 4, 5, 6], 'Reception'); // Очко соперника - наш приём
        this.setRuleForActionAndQuality('OpponentError', [1, 2, 3, 4, 5, 6], 'Serve'); // Ошибка соперника - наша подача
        
        // Win/Loss Rules - правила результата сегмента
        this.setDefaultWinLoseLogic();
    }

    /**
     * Установить правила результата по умолчанию
     */
    setDefaultWinLoseLogic() {
        // Результаты: NotEnded = 0, Won = 1, Lost = 2, Disputable = 3
        // Индексы массива: quality 1-6
        
        this.setWinLoseRule('Serve', [2, 0, 0, 0, 0, 1]); // Q1=Lost, Q2-5=NotEnded, Q6=Won
        this.setWinLoseRule('Reception', [2, 0, 0, 0, 0, 0]); // Q1=Lost, остальные NotEnded
        this.setWinLoseRule('Set', [2, 0, 0, 0, 0, 0]); // Q1=Lost, остальные NotEnded
        this.setWinLoseRule('Attack', [2, 2, 0, 0, 0, 1]); // Q1-2=Lost, Q3-5=NotEnded, Q6=Won
        this.setWinLoseRule('Block', [2, 0, 0, 0, 0, 1]); // Q1=Lost, Q2-5=NotEnded, Q6=Won
        this.setWinLoseRule('Defence', [2, 0, 0, 0, 0, 0]); // Q1=Lost, остальные NotEnded
        this.setWinLoseRule('FreeBall', [2, 0, 0, 0, 0, 0]); // Q1=Lost, остальные NotEnded
        this.setWinLoseRule('Transfer', [2, 0, 0, 0, 0, 0]); // Q1=Lost, остальные NotEnded
        this.setWinLoseRule('OpponentError', [1, 1, 1, 1, 1, 1]); // Всегда Won
        this.setWinLoseRule('OpponentPoint', [2, 2, 2, 2, 2, 2]); // Всегда Lost
    }

    /**
     * Получить возможные следующие действия
     * @param {string} actionType - тип действия
     * @param {number} quality - качество (1-6)
     * @returns {Array<string>} массив возможных действий
     */
    getPossibleActions(actionType, quality) {
        if (!this.sequenceLogic[actionType]) {
            return [];
        }
        return this.sequenceLogic[actionType][quality - 1] || [];
    }

    /**
     * Получить возможные действия для соперника
     * @param {string} actionType - тип действия соперника
     * @returns {Array<string>} массив возможных действий
     */
    getPossibleActionsForOpponent(actionType) {
        // Для соперника используем quality = 2 (из C# кода - index 1)
        if (!this.sequenceLogic[actionType]) {
            return [];
        }
        return this.sequenceLogic[actionType][1] || [];
    }

    /**
     * Получить результат действия
     * @param {string} actionType - тип действия
     * @param {number} quality - качество (1-6)
     * @returns {number} результат (0=NotEnded, 1=Won, 2=Lost, 3=Disputable)
     */
    getActionResult(actionType, quality) {
        if (!this.winLossLogic[actionType]) {
            return 0; // NotEnded
        }
        return this.winLossLogic[actionType][quality - 1];
    }

    /**
     * Получить результат для действия соперника
     * @param {string} actionType - тип действия соперника
     * @returns {number} результат (1=Won, 2=Lost)
     */
    getActionResultForOpponent(actionType) {
        if (actionType === 'OpponentError') {
            return 1; // Won
        }
        return 2; // Lost
    }

    /**
     * Получить результат для действия судьи
     * @param {string} actionType - тип действия судьи
     * @returns {number} результат
     */
    getActionResultForJudge(actionType) {
        if (actionType === 'DisputableBall') {
            return 3; // Disputable
        }
        if (actionType === 'JudgeMistakeLost') {
            return 2; // Lost
        }
        return 1; // Won (JudgeMistakeWon)
    }

    /**
     * Получить начальные действия игроков для старта сета
     * @returns {Array<string>} массив действий
     */
    getSetStartPlayersActions() {
        return ['Serve', 'Reception'];
    }

    /**
     * Получить действия приёма
     * @returns {Array<string>} массив действий
     */
    getReceptionPlayersActions() {
        return ['Reception'];
    }
}
